File: bindings/python/examples/kv.py
"""kv: a wires-native service written in Python (card 33).

The same store as the Rust example (wires/examples/kv.rs): a key-value map
held in this process's memory, one namespace per verified person, so each
caller sees only their own keys.

    wires call kv -- set greeting <<< 'hello'   # the value is stdin
    wires call kv -- get greeting               # hello
    wires call kv -- keys                       # greeting

Run it from a joined node's keystore, trusting one IdP:

    ./.scripts/build-python.sh
    PYTHONPATH=target/python python3 bindings/python/examples/kv.py \\
        <WIRES_HOME> <issuer> <audience> [--push-to ROLE] [--loopback]

With --push-to, a `set` also pushes "kv: <key> set" to the caller's
`wires inbox` (members of ROLE may receive pushes). With --loopback, the
host takes direct connections only from this machine (others come through
its relay), so the macOS firewall doesn't prompt for Python.
"""

import argparse
import sys
import threading

import wires

MAX_VALUE = 1024 * 1024


class Kv(wires.Service):
    def __init__(self, push: bool):
        self._push = push
        self._lock = threading.Lock()
        self._people: dict[tuple[str, str], dict[str, bytes]] = {}

    def call(self, call: wires.Call) -> int:
        person = call.principal()
        if person is None:
            call.write_stderr(b"kv: no verified identity\n")
            return 1
        me = (person.issuer, person.subject)
        match call.args():
            case ["set", key]:
                value = call.read_all_stdin()
                if len(value) > MAX_VALUE:
                    call.write_stderr(b"kv: value over 1 MiB\n")
                    return 1
                with self._lock:
                    self._people.setdefault(me, {})[key] = value
                if self._push:
                    call.push_to_caller(f"kv: {key} set", f"{len(value)} bytes")
                return 0
            case ["get", key]:
                with self._lock:
                    value = self._people.get(me, {}).get(key)
                if value is None:
                    call.write_stderr(b"kv: no such key\n")
                    return 1
                call.write_stdout(value)
                return 0
            case ["keys"]:
                with self._lock:
                    keys = sorted(self._people.get(me, {}))
                call.write_stdout("".join(f"{k}\n" for k in keys).encode())
                return 0
            case _:
                call.write_stderr(b"usage: kv set KEY (value on stdin) | get KEY | keys\n")
                return 2


def main() -> int:
    parser = argparse.ArgumentParser(description="kv: a wires-native service in Python")
    parser.add_argument("home", help="a joined node's keystore (WIRES_HOME)")
    parser.add_argument("issuer", help="the IdP whose ID tokens to trust")
    parser.add_argument("audience", help="the OAuth client id tokens are issued to")
    parser.add_argument("--push-to", metavar="ROLE", help="let `set` push to members of ROLE")
    parser.add_argument("--loopback", action="store_true", help="direct connections from this machine only")
    args = parser.parse_args()
    push_to = args.push_to
    builder = wires.HostBuilder(args.home).trust_issuer(args.issuer, [args.audience])
    if push_to:
        builder = builder.push_allow([push_to])
    if args.loopback:
        builder = builder.bind_loopback()
    host = builder.service("kv", Kv(push=push_to is not None)).build()
    print(f"kv: serving as {host.node_id()}", file=sys.stderr, flush=True)
    host.serve()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: wires/examples/kv.rs
//! `kv`: a wires-native service (card 33). A key-value store held in the
//! daemon's memory, with one namespace per verified person, so each caller
//! sees only their own keys. A CLI wrapper couldn't do this cheaply: the
//! state lives across calls in one warm process, and the caller's identity
//! arrives as a type, not an environment variable to parse.
//!
//! ```text
//! wires call kv -- set greeting <<< 'hello'   # the value is stdin
//! wires call kv -- get greeting               # hello
//! wires call kv -- keys                       # greeting
//! ```
//!
//! Run it from a joined node's keystore, trusting one IdP:
//!
//! ```text
//! cargo run -p wires --example kv -- <WIRES_HOME> <issuer> <audience>
//! ```
//!
//! The admin must have registered `kv` on this node (`wires service add kv
//! --allow <role> --host <this node>`), or the host refuses to start.

use std::collections::{BTreeMap, HashMap};
use std::sync::Mutex;

use tokio::io::{AsyncReadExt, AsyncWriteExt};

/// A verified person: their IdP's issuer and subject.
type Person = (String, String);

/// One person's keys and values.
type Keys = BTreeMap<String, Vec<u8>>;

/// The store: each person's own keys.
#[derive(Default)]
pub struct Kv {
    people: Mutex<HashMap<Person, Keys>>,
}

/// The largest value `set` keeps.
const MAX_VALUE: u64 = 1024 * 1024;

impl wires::Service for Kv {
    async fn call(&self, call: wires::Call, mut io: wires::CallIo) -> i32 {
        let Some(person) = call.principal() else {
            let _ = io.stderr.write_all(b"kv: no verified identity\n").await;
            return 1;
        };
        let person = (person.issuer.clone(), person.subject.clone());
        let args: Vec<&str> = call.args().iter().map(String::as_str).collect();
        match args.as_slice() {
            ["set", key] => {
                let mut value = Vec::new();
                let read = (&mut io.stdin)
                    .take(MAX_VALUE + 1)
                    .read_to_end(&mut value)
                    .await;
                if read.is_err() || value.len() as u64 > MAX_VALUE {
                    let _ = io
                        .stderr
                        .write_all(b"kv: value unreadable or over 1 MiB\n")
                        .await;
                    return 1;
                }
                let mut people = self.people.lock().expect("kv poisoned");
                people
                    .entry(person)
                    .or_default()
                    .insert(key.to_string(), value);
                0
            }
            ["get", key] => {
                let value = {
                    let people = self.people.lock().expect("kv poisoned");
                    people.get(&person).and_then(|keys| keys.get(*key)).cloned()
                };
                match value {
                    Some(value) => io.stdout.write_all(&value).await.map_or(1, |()| 0),
                    None => {
                        let _ = io.stderr.write_all(b"kv: no such key\n").await;
                        1
                    }
                }
            }
            ["keys"] => {
                let listing: String = {
                    let people = self.people.lock().expect("kv poisoned");
                    people
                        .get(&person)
                        .map(|keys| keys.keys().map(|k| format!("{k}\n")).collect())
                        .unwrap_or_default()
                };
                io.stdout
                    .write_all(listing.as_bytes())
                    .await
                    .map_or(1, |()| 0)
            }
            _ => {
                let _ = io
                    .stderr
                    .write_all(b"usage: kv set KEY (value on stdin) | get KEY | keys\n")
                    .await;
                2
            }
        }
    }
}

#[allow(dead_code)] // the e2e tests include this file for `Kv` alone
#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let [home, issuer, audience] = args.as_slice() else {
        anyhow::bail!("usage: kv <WIRES_HOME> <issuer> <audience>");
    };
    let host = wires::Host::builder(home)
        .trust_issuer(issuer, [audience])
        .service("kv", Kv::default())
        .build()?;
    eprintln!("kv: serving as {}", host.node_id().hex());
    host.serve().await
}
